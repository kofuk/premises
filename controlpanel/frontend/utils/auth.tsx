import {ReactNode, createContext, useContext} from 'react';

import {login as apiLogin, useSessionData} from '@/api';
import Loading from '@/components/loading';

export enum LoginResult {
  LoggedIn,
  NeedsChangePassword
}

type AuthContextType = {
  loggedIn: boolean;
  userName: string | null;
  login: (userName: string, password: string) => Promise<LoginResult>;
  logout: () => Promise<void>;
  initializePassword: (username: string, newPassword: string) => Promise<void>;
};

const AuthContext = createContext<AuthContextType>(null!);

export const AuthProvider = ({children}: {children: ReactNode}) => {
  const {data: session, error, isLoading, mutate} = useSessionData();
  if (isLoading) {
    return <Loading />;
  }
  if (error) {
    throw error;
  }

  const login = async (userName: string, password: string): Promise<LoginResult> => {
    const resp = await apiLogin({userName, password});
    if (resp.needsChangePassword) {
      return LoginResult.NeedsChangePassword;
    }

    mutate();

    return LoginResult.LoggedIn;
  };

  const logout = async () => {
    const resp = await fetch('/logout', {method: 'POST'}).then((resp) => resp.json());
    if (!resp['success']) {
      throw new Error(resp['reason']);
    }
    mutate();
  };

  const initializePassword = async (username: string, newPassword: string): Promise<void> => {
    const params = new URLSearchParams();
    params.append('username', username);
    params.append('password', newPassword);

    const resp = await fetch('/login/reset-password', {
      method: 'post',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded'
      },
      body: params.toString()
    }).then((resp) => resp.json());

    if (!resp['success']) {
      throw new Error(resp['reason']);
    }

    mutate();
  };

  const value = {
    loggedIn: !!session && session.loggedIn,
    userName: (session && session.userName) || null,
    login,
    logout,
    initializePassword
  };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export const useAuth = () => {
  return useContext(AuthContext);
};
