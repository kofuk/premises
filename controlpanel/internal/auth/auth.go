package auth

import (
	"context"
	"encoding/base32"
	"time"

	"github.com/gorilla/securecookie"
	"github.com/kofuk/premises/controlpanel/internal/kvs"
)

type Scope string

const (
	NoScope    Scope = ""
	AdminScope Scope = "admin"
)

type AuthService struct {
	kvs kvs.KeyValueStore
}

type Token struct {
	Token     string    `json:"token"`
	UserID    uint      `json:"userID"`
	Scopes    []Scope   `json:"scopes"`
	CreatedAt time.Time `json:"createdAt"`
}

func New(kvs kvs.KeyValueStore) *AuthService {
	return &AuthService{
		kvs: kvs,
	}
}

func (a *AuthService) CreateToken(ctx context.Context, userID uint, scopes []Scope) (*Token, error) {
	token := &Token{
		Token:     base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(securecookie.GenerateRandomKey(32)),
		UserID:    userID,
		Scopes:    scopes,
		CreatedAt: time.Now(),
	}

	if err := a.kvs.Set(ctx, "token:"+token.Token, token, 30*24*time.Hour); err != nil {
		return nil, err
	}

	return token, nil
}

func (a *AuthService) Get(ctx context.Context, token string) (*Token, error) {
	var t Token
	if err := a.kvs.Get(ctx, "token:"+token, &t); err != nil {
		return nil, err
	}

	return &t, nil
}

func (a *AuthService) RevokeToken(ctx context.Context, token string) error {
	return a.kvs.Del(ctx, "token:"+token)
}

func (a *AuthService) IsGranted(ctx context.Context, token string, scope Scope) (bool, error) {
	var t Token
	if err := a.kvs.Get(ctx, "token:"+token, &t); err != nil {
		return false, err
	}

	if scope == NoScope {
		return true, nil
	}

	for _, s := range t.Scopes {
		if s == scope {
			return true, nil
		}
	}

	return false, nil
}
