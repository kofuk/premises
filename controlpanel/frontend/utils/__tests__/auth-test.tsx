import React from 'react';

import {render} from '@testing-library/react';
import nock from 'nock';
import fetch from 'node-fetch';

import {AuthProvider} from '../auth';

const InnerComponent = () => {
  return <div>Test</div>;
};

describe('AuthProvider', () => {
  beforeEach(() => {
    location.replace(`http://localhost`);
    global.fetch = fetch as any as typeof global.fetch;
  });

  it('renders component after initialized', () => {
    nock('http://localhost').get('/api/current-user').reply(200, {
      success: true,
      user_name: 'hoge'
    });

    const component = render(
      <AuthProvider>
        <InnerComponent />
      </AuthProvider>
    );
    const tree = component.container;
    expect(tree).toMatchSnapshot();
  });
});
