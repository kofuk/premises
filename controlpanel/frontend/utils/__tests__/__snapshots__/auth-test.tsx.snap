// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthProvider renders component after initialized 1`] = `
<div>
  <div
    class="MuiBox-root css-f6a0yb"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-ayl0v9-MuiSkeleton-root"
      style="height: 200px;"
    />
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-ayl0v9-MuiSkeleton-root"
      style="height: 40px;"
    />
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-ayl0v9-MuiSkeleton-root"
      style="width: 80%; height: 40px;"
    />
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave css-ayl0v9-MuiSkeleton-root"
      style="width: 95%; height: 40px;"
    />
  </div>
</div>
`;
